### PR TITLE
Adds start and stop work heartbeats.

### LINKED ISSUE #1186: StartWork / StopWork pulse from Overseer based on block import events
#1152 implemented the subsystem-management aspect of the overseer, but not the `StartWork`/`StopWork` heartbeat described in the doc. This heartbeat is necessary to implement subsystems

### PR DESCRIPTION
**Pull request by montekki**
Adds the `StartWork` and `StopWork` heartbeats to the `Overseer`. The initial set of leaves is passed to the `Overseer` via a parameter in constructor, my assumption is that the necessary set of leaves is "lifted" upon startup here (and other calls to `leaves()`:

https://github.com/paritytech/polkadot/blob/162486cd20175012aaebce71a4df3d78efdc8ec5/service/src/lib.rs#L328-L330

That's why I decided to pass them as a parameter instead of making `Overseer` generic over some backend.
Block importing is done only based on block `Hash` and  parent `Hash`, no dependency on `Header` here.

So atm code only depends on the `Hash` from outside world. 

Fixes #1186

### CODE REVIEW

**Comment by montekki** [comment]
The guide doesn't seem to mention it but probably `StopWork` has to be sent on other conditions as well (not only for parents of imported blocks) so we can kill work on orphaned blocks.

**Comment by rphmeier** [comment]
Ah, good point. That's why we wanted to handle block finalization events, but at the time I was drafting that section of the guide it slipped my mind.

What we want is for all block finalization events to lead to `StopWork` of all blocks with that number, so removing them from the active leaves set. I'll prepare a PR to add that information to the guide.

File: overseer/src/lib.rs [hunk 1]
```diff
@@ -537,6 +564,13 @@ where
 
 	/// Run the `Overseer`.
 	pub async fn run(mut self) -> SubsystemResult<()> {
+		let leaves = std::mem::replace(&mut self.leaves, vec![]);
+
+		for leaf in leaves.into_iter() {
```
Comment (rphmeier) [inline overseer/src/lib.rs:569]: `leaves.drain(..)` would make the above line redundant

Comment (montekki) [reply overseer/src/lib.rs:569]: look like that would need two mutable refs to `self`?

Comment (rphmeier) [reply overseer/src/lib.rs:569]: Does this work?

```rust
for leaf in self.leaves.drain(..) {

}

// borrow of `self` should not last here, but if rustc thinks it does, we can wrap in an enclosing scope
```

File: overseer/src/lib.rs [hunk 2]
```diff
@@ -471,6 +489,7 @@ where
 	/// # }); }
 	/// ```
 	pub fn new(
+		leaves: &[Hash],
```
Comment (bkchr) [inline overseer/src/lib.rs:492]:
```suggestion
		leaves: impl Into<Vec<Hash>>,
```

You take a slice and convert it into a `Vec` anyway. This makes the user either pass a `Vec` directly or still pass the slice.

File: overseer/src/lib.rs [hunk 3]
```diff
@@ -537,6 +564,13 @@ where
 
 	/// Run the `Overseer`.
 	pub async fn run(mut self) -> SubsystemResult<()> {
+		let leaves = std::mem::replace(&mut self.leaves, vec![]);
```
Comment (bkchr) [inline overseer/src/lib.rs:567]:
```suggestion
		let leaves = std::mem::take(&mut self.leaves);
```

File: overseer/src/lib.rs (hunk 1 above)
Comment (montekki) [reply overseer/src/lib.rs:569]:
```rust
		for leaf in self.leaves.drain(..) {
			self.broadcast_signal(OverseerSignal::StartWork(leaf)).await?;
			self.active_leaves.insert(leaf);
		}

```

these are two mutable refs: the `drain` is the first mutable refs and calls to `self` in the body of the loop are other mutable refs.

Comment (rphmeier) [reply overseer/src/lib.rs:567]: I see. Then Basti's suggestion of `mem::take` makes most sense. That way we will deallocate the underlying storage, unlike in my suggestion.

File: overseer/src/lib.rs [hunk 4]
```diff
@@ -498,7 +498,7 @@ where
 	/// # }); }
 	/// ```
 	pub fn new(
-		leaves: impl Into<Vec<BlockInfo>>,
+		leaves: impl IntoIterator<Item = BlockInfo>,
```
Comment (rphmeier) [inline overseer/src/lib.rs:501]: 🙏

**Pull request merged**